UI to block case list/detail translations to go to transifex

emord opened the pull request:
@mkangia This is a UI to block case lists and details. I was originally just going to use an admin page and that got lost in a merge conflict a while ago and wasn't noticed.

I added this to your existing translation management page as that seemed like the most obvious place to put it for now.

![image](https://user-images.githubusercontent.com/1471773/53446747-a9c20380-39e1-11e9-847a-543d883ec1aa.png)

Not bothering with paginating, because this table shouldn't get too large, roughly ~30 or so now.

emord committed:
Add TransifexBlacklist to admin page

emord committed:
List translations that are blacklisted

emord committed:
Translate app_id to app name

emord committed:
Add to transifex blacklist from app translations page

emord committed:
Flip table and add new translation page

emord commented on corehq/apps/translations/integrations/transifex/views.py:
```diff
@@ -326,11 +329,24 @@ def page_context(self):
+        ret = []
+        for trans in blacklisted:
+            r = trans.copy()
+            r['app_name'] = app_ids_to_name.get(trans['app_id'], trans['app_id'])
```
I tried looking to see if I could easily get the module name instead of id, but I think that would require pulling all the apps that are reference to find that. There's no easy view to get that info.

orangejenny replied:
makes sense to me

orangejenny commented on corehq/apps/translations/forms.py:
```diff
@@ -258,3 +260,41 @@ class TransifexOrganizationForm(forms.ModelForm):
+
+    def __init__(self, domain, *args, **kwargs):
+        super(AddTransifexBlacklistForm, self).__init__(*args, **kwargs)
+        self.helper = FormHelper()
```
Use `HQFormHelper` here and drop the `label_class` and `field_class`, unless there's a reason for this form to be unique.

orangejenny commented on corehq/apps/translations/templates/app_translations.html:
```diff
@@ -52,6 +53,32 @@
+            <div id="blacklist-trans-form" class="tab-pane">
+                <table class="table table-striped table-hover">
+                    <thead>
+                        <th class="col-sm-2">App</th>
```
translations?

orangejenny commented on corehq/apps/translations/templates/app_translations.html:
```diff
@@ -52,6 +53,32 @@
+                    {% endfor %}
+                    </tbody>
+                </table>
+                <form class="form form-horizontal" name="product" method="post">
```
I don't think you need this - `{% crispy blacklist_form %}` ought to include the form tag, and you can use a `Fieldset` to make the legend.

orangejenny commented on corehq/apps/translations/forms.py:
```diff
@@ -258,3 +260,41 @@ class TransifexOrganizationForm(forms.ModelForm):
+        form_fields.append(hqcrispy.Field(StrictButton(
+            ugettext_lazy("Add"),
+            type="submit",
+            css_class="btn btn-primary btn-lg disable-on-submit",
```
Why is this `btn-lg`?

orangejenny commented on corehq/apps/translations/forms.py:
```diff
@@ -258,3 +260,41 @@ class TransifexOrganizationForm(forms.ModelForm):
+            hqcrispy.Field('domain'),
+            hqcrispy.Field('action'),
+        ]
+        form_fields.append(hqcrispy.Field(StrictButton(
```
Use `FormActions` instead of the single button.

mkangia commented:
I would push for not adding this to the "Manage App Translations" page as that is focussed on the process of communicating with transifex for application translations.
This should be added as a separate page like "Convert Translations"/ "Pull resources".

mkangia commented on corehq/apps/translations/templates/app_translations.html:
```diff
@@ -52,6 +53,32 @@
+                    {% for translation in blacklisted_translations %}
+                        <tr>
+                            <td>{{ translation.app_name }}</td>
+                            <td>{{ translation.module_id }}</td>
```
Can we make this a link to the module? So that one can see what module this is coz the name is not present here.

mkangia replied:
Can also reconsider this if decide to put this out as a separate page.

emord committed:
Add translations

emord committed:
Merge branch 'master' into je/transifex-blacklist-ui

emord committed:
Move translation blacklist form

emord committed:
Change form to use HQFormHelper

emord committed:
Add link to module

emord committed:
Remove blacklist form from manage app translation form

emord committed:
Remove btn-lg

emord committed:
Use FormAction

emord committed:
Fix form saving

emord commented:
@orangejenny Thanks for the detailed feedback. I think I fixed all of it. Is there a source of truth on how to setup a form in HQ? I found https://commcare-hq.readthedocs.io/forms.html https://www.commcarehq.org/styleguide/docs/simple_crispy/forms/ and https://www.commcarehq.org/styleguide/molecules/#molecules-checkboxes which all used `FormHelper`

@mkangia pushed it to a different page and added the link

mkangia commented on corehq/apps/translations/integrations/transifex/views.py:
```diff
@@ -296,6 +299,47 @@ def post(self, request, *args, **kwargs):
+    @property
+    def _blacklisted_translations(self):
+        blacklisted = TransifexBlacklist.objects.filter(domain=self.domain).all().values()
+        app_ids_to_name = {app.id: app.name for app in get_brief_apps_in_domain(self.domain)}
```
i think you can add a method like [to_json](https://github.com/dimagi/commcare-hq/blob/9b750c2226b88bbeacd7238df3feb97e340208a9/corehq/apps/locations/models.py#L511) to `TransifexBlacklist` that does this conversion for you. I believe adding such parsing on the model is a better thing than adding it here so that it can be reused. There might be even better ways to add a parser(rails calls it presenter) to just handle this logic and not even model if you want to push for it.

mkangia replied:
i meant to write for the logic that follows now to get the app_name in.

emord replied:
https://github.com/dimagi/commcare-hq/pull/23429/commits/46627206cb30fa1144351950ccfbbbd71c033af2

mkangia commented on corehq/tabs/tabclasses.py:
```diff
@@ -1349,12 +1349,18 @@ def sidebar_items(self):
+                    {
+                        'url': reverse('blacklist_translations', args=[self.domain]),
+                        'title': 'Blacklist Translations'
+                    },
```
👍

mkangia commented on corehq/apps/translations/forms.py:
```diff
@@ -271,3 +271,43 @@ class TransifexOrganizationForm(forms.ModelForm):
+        self.fields['app_id'].choices = tuple((app.id, app.name) for app in get_brief_apps_in_domain(domain))
+        form_fields = [
+            hqcrispy.Field('app_id'),
+            hqcrispy.Field('module_id'),
```
you dont need to add `module_id` and couple of others on the form as properties above?

Should this form validate things like
1. module_id passed is actually a module in the app?
2. maybe, display_text/field name is actually present in the module?

emord replied:
https://github.com/dimagi/commcare-hq/pull/23429/commits/60a03f62238db5a87a26358ab8c69bd66a2fcd17

> maybe, display_text/field name is actually present in the module?
 I think that's a bit harder as ICDS/REACH uses some custom case list XML. We don't have the ability to parse that at the moment (which is why its custom) so we wouldn't be able to confidently validate that

emord replied:
> you dont need to add module_id and couple of others on the form as properties above?

under the `Meta` class, it has `fields = __all__` You only need to add at the class level if you want to override them

orangejenny commented on corehq/apps/translations/forms.py:
```diff
@@ -303,7 +300,10 @@ def __init__(self, domain, *args, **kwargs):
         self.helper.layout = crispy.Layout(
-            *form_fields
+            crispy.Fieldset(
+                "Add translation to blacklist",
```
translate?

orangejenny commented:
@emord Thanks for updates and for looking at the docs - I had forgotten about forms.rst and apparently didn't fully update the style guide example. Updating/reconciling docs in https://github.com/dimagi/commcare-hq/pull/23671

emord committed:
Add translations

emord committed:
Move app name retrieval to TransifexBlacklist

emord committed:
Increase size limit for app and module ids.

Module and app ids aren't UUIDs as I originally thought. They are 40
character hex values.

Note that generally changing these constraints is not a good idea, but
this table is not used yet, so it is safe

emord committed:
Add validation that module exists in app

emord committed:
Fix help text to include HTML

emord committed:
Add absolute_import

emord commented on corehq/apps/translations/models.py:
```diff
@@ -129,8 +129,8 @@ class TransifexBlacklist(models.Model):
-    app_id = models.CharField(max_length=32)
-    module_id = models.CharField(max_length=32)
+    app_id = models.CharField(max_length=255)
+    module_id = models.CharField(max_length=255)
```
I found out while testing that these aren't actually UUIDs, and are actually 40 characters for some reason. I'll be following up on a separate PR to change that

emord committed:
Retrieve app name from already loaded json

mkangia commented on corehq/apps/translations/integrations/transifex/views.py:
```diff
@@ -296,6 +298,36 @@ def post(self, request, *args, **kwargs):
+        if self.transifex_integration_enabled(request):
+            if self.blacklist_form.is_valid():
+                self.blacklist_form.save()
+        return self.get(request, *args, **kwargs)
```
i think this would result in the form being filled with previous values even in case of success save. A redirect here in case of success would refresh everything.

emord replied:
When I was testing this out, I kind of liked it since it would have the app/module filled out which would make it easier to fill out several additions at once. I will wait for GS feedback to decide if they do not want that
